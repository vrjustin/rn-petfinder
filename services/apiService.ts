import axios, {AxiosResponse, AxiosError} from 'axios';
import PetType from '../models/PetType';
import Breed from '../models/Breed';
import Animal from '../models/Animal';

interface AccessTokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
}

interface PetTypesResponse {
  types: PetType[];
}

let jwt_access_token: string | null = null;

const AUTH_ENDPOINT = 'https://api.petfinder.com/v2/oauth2/token';
const TYPES_ENDPOINT = 'https://api.petfinder.com/v2/types';

const getAccessToken = async (): Promise<string | null> => {
  try {
    const response: AxiosResponse<AccessTokenResponse> = await axios.post(
      AUTH_ENDPOINT,
      {
        grant_type: 'client_credentials',
        client_id: process.env.CLIENT_ID!,
        client_secret: process.env.CLIENT_SECRET!,
      },
    );

    const {access_token} = response.data;
    jwt_access_token = access_token;
    return access_token;
  } catch (error) {
    console.log('-----------');
    console.log(error);
    console.log('-----------');
    console.error(error);
    return null;
  }
};

const getPetTypes = async (): Promise<PetTypesResponse> => {
  if (!jwt_access_token) {
    await getAccessToken();
  }

  try {
    const response: AxiosResponse<PetTypesResponse> = await axios.get(
      TYPES_ENDPOINT,
      {
        headers: {
          Authorization: `Bearer ${jwt_access_token}`,
        },
      },
    );

    return response.data;
  } catch (error) {
    console.error('Failed to get pet types:', error);
    return {types: []};
  }
};

const getPetBreeds = async (type: string): Promise<Breed[]> => {
  if (!jwt_access_token) {
    await getAccessToken();
  }
  try {
    const response = await axios.get(
      `https://api.petfinder.com/v2/types/${type}/breeds`,
      {
        headers: {
          Authorization: `Bearer ${jwt_access_token}`,
        },
      },
    );

    return response.data.breeds;
  } catch (error) {
    console.error('Failed to fetch pet breeds:', error);
    return [];
  }
};

const getAnimals = async (type: string, breed: string): Promise<Animal[]> => {
  if (!jwt_access_token) {
    await getAccessToken();
  }
  try {
    const response = await axios.get(
      `https://api.petfinder.com/v2/animals?type=${type}&breed=${breed}`,
      {
        headers: {
          Authorization: `Bearer ${jwt_access_token}`,
        },
      },
    );

    return response.data.animals;
  } catch (error: unknown) {
    console.error('Failed to fetch animals from type & breed:', error);
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (axiosError.response?.status === 401) {
        console.log('Received 401 error, refreshing token...');
        await getAccessToken();
        return getAnimals(type, breed);
      }
    }
    console.error('Failed to fetch animals from type & breed: ', error);
    return [];
  }
};

export default {
  getAccessToken,
  getPetTypes,
  getPetBreeds,
  getAnimals,
};
