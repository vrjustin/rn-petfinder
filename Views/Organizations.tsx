import React, {useEffect, useState} from 'react';
import {
  View,
  Text,
  FlatList,
  TouchableOpacity,
  StyleSheet,
  Linking,
} from 'react-native';
import GlobalStyles from './Styles/GlobalStyles';
import {useSelector, useDispatch} from 'react-redux';
import {
  setOrganizations,
  selectOrganizations,
} from '../reducers/organizationsReducer';
import {selectSearchParameters} from '../reducers/searchParamsReducer';
import apiService from '../services/apiService';
import Organization from '../models/Organization';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';

const Organizations = () => {
  const dispatch = useDispatch();
  const orgs = useSelector(selectOrganizations);
  const globalStyles = GlobalStyles();
  const searchParameters = useSelector(selectSearchParameters);
  const {location, distance} = searchParameters;
  const [currentPage, setCurrentPage] = useState(1);
  const [totalPages, setTotalPages] = useState(1);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const fetchOrganizations = async () => {
      try {
        setIsLoading(true);
        const orgsResponse = await apiService.getOrganizations(
          location.zipCode,
          distance,
          currentPage,
        );
        const {organizations, pagination} = orgsResponse;
        dispatch(setOrganizations(organizations));
        setCurrentPage(pagination.current_page);
        setTotalPages(pagination.total_pages);
      } catch (error) {
        console.error('Failed to fetch organizations: ', error);
      } finally {
        setIsLoading(false);
      }
    };
    fetchOrganizations();
  }, [dispatch, distance, location, currentPage]);

  const renderOrg = ({item}: {item: Organization}) => (
    <TouchableOpacity
      style={styles.orgContainer}
      onPress={() => Linking.openURL(item.url)}>
      <Text style={styles.orgName}>{item.name}</Text>
      <Text
        style={
          styles.orgLocation
        }>{`${item.address.city}, ${item.address.state}`}</Text>
      <Text style={styles.orgHours}>{renderHours(item.hours)}</Text>
    </TouchableOpacity>
  );

  const renderHours = (hours: Organization['hours']) => {
    const days = [
      'Monday',
      'Tuesday',
      'Wednesday',
      'Thursday',
      'Friday',
      'Saturday',
      'Sunday',
    ];
    return days
      .map(day => {
        const hour = hours[day.toLowerCase()];
        return hour ? `${day}: ${hour}` : null;
      })
      .filter(Boolean)
      .join('\n');
  };

  const prevPage = () => {
    if (isLoading) {
      return;
    }
    setCurrentPage(prev => Math.max(prev - 1, 1));
  };

  const nextPage = () => {
    if (isLoading) {
      return;
    }
    setCurrentPage(prev => Math.min(prev + 1, totalPages));
  };

  const paginationHeader = () => {
    return (
      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'center',
          height: 32,
        }}>
        <TouchableOpacity onPress={prevPage}>
          <FontAwesomeIcon name="caret-left" size={30} style={styles.icon} />
        </TouchableOpacity>
        <Text style={{marginHorizontal: 16}}>
          {currentPage} / {totalPages}
        </Text>
        <TouchableOpacity onPress={nextPage}>
          <FontAwesomeIcon name="caret-right" size={30} style={styles.icon} />
        </TouchableOpacity>
      </View>
    );
  };

  return (
    <View style={globalStyles.container}>
      {totalPages > 1 ? paginationHeader() : <></>}
      <FlatList
        data={orgs}
        renderItem={renderOrg}
        keyExtractor={org => org.id}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  orgContainer: {
    backgroundColor: '#f9f9f9',
    padding: 16,
    marginBottom: 8,
    borderRadius: 8,
  },
  orgName: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 4,
  },
  orgLocation: {
    fontSize: 16,
  },
  orgHours: {
    fontSize: 14,
    color: 'gray',
  },
  icon: {
    marginRight: 8,
  },
});

export default Organizations;
