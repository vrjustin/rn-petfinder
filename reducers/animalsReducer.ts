import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Animal from '../models/Animal';

interface AnimalsState {
  animals: Animal[];
  favorites: Animal[];
}

const initialState: AnimalsState = {
  animals: [],
  favorites: [],
};

const animalsSlice = createSlice({
  name: 'animals',
  initialState,
  reducers: {
    setAnimals(state, action: PayloadAction<Animal[]>) {
      state.animals = action.payload;
    },
    toggleFavorite(state, action: PayloadAction<Animal>) {
      const isFavorite = state.favorites.some(
        fav => fav.id === action.payload.id,
      );
      if (isFavorite) {
        state.favorites = state.favorites.filter(
          fav => fav.id !== action.payload.id,
        );
      } else {
        state.favorites.push(action.payload);
      }
    },
  },
});

export const selectAnimals = (state: RootState) => state.animals.animals;
export const selectFavorites = (state: RootState) => state.animals.favorites;

export const {setAnimals, toggleFavorite} = animalsSlice.actions;
export default animalsSlice.reducer;
