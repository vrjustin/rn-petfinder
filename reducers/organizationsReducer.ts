import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Organization from '../models/Organization';

interface OrganizationsState {
  organizations: Organization[];
}

const initialState: OrganizationsState = {
  organizations: [],
};

const organizationsSlice = createSlice({
  name: 'organizations',
  initialState,
  reducers: {
    setOrganizations(state, action: PayloadAction<Organization[]>) {
      state.organizations = action.payload;
    },
  },
});

export const selectOrganizations = (state: RootState) =>
  state.organizations.organizations;

export const {setOrganizations} = organizationsSlice.actions;
export default organizationsSlice.reducer;
