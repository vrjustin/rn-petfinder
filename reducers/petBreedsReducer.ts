import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import Breed from '../models/Breed';
import {RootState} from '../stores/store';

interface BreedsState {
  breeds: Breed[];
}

const initialState: BreedsState = {
  breeds: [],
};

const breedsSlice = createSlice({
  name: 'breeds',
  initialState,
  reducers: {
    setBreeds(state, action: PayloadAction<Breed[]>) {
      state.breeds = action.payload;
    },
  },
});

export const selectPetBreeds = (state: RootState) => state.petBreeds.breeds;

export const {setBreeds} = breedsSlice.actions;
export default breedsSlice.reducer;
