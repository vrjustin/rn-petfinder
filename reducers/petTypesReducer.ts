import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import {PetType} from '../models/PetType';

interface PetTypesState {
  petTypes: PetType[];
}

const initialState: PetTypesState = {
  petTypes: [],
};

const petTypesSlice = createSlice({
  name: 'petTypes',
  initialState,
  reducers: {
    setPetTypes(state, action: PayloadAction<PetType[]>) {
      state.petTypes = action.payload;
    },
  },
});

export const selectPetTypes = (state: RootState) => state.petTypes.petTypes;

export const {setPetTypes} = petTypesSlice.actions;
export default petTypesSlice.reducer;
