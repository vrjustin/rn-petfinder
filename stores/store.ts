import {combineReducers, configureStore} from '@reduxjs/toolkit';
import petTypesReducer from '../reducers/petTypesReducer';
import petBreedsReducer from '../reducers/petBreedsReducer';
import animalsReducer from '../reducers/animalsReducer';
import searchParamsReducer from '../reducers/searchParamsReducer';

const createDebugger = require('redux-flipper').default;

const rootReducer = combineReducers({
  petTypes: petTypesReducer,
  petBreeds: petBreedsReducer,
  animals: animalsReducer,
  searchParameters: searchParamsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    __DEV__
      ? getDefaultMiddleware({serializableCheck: false}).concat(
          createDebugger(),
        )
      : getDefaultMiddleware({serializableCheck: false}),
});

export default store;
