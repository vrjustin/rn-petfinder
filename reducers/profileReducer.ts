import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Profile from '../models/Profile';

interface ProfileState {
  profile: Profile;
}

const initialState: ProfileState = {
  profile: {
    shouldOnboard: true,
    isRehydrated: false,
  },
};

const profileSlice = createSlice({
  name: 'profile',
  initialState,
  reducers: {
    setProfile(state, action: PayloadAction<Profile>) {
      state.profile = action.payload;
    },
  },
});

export const profile = (state: RootState) => state.profile.profile;

export const {setProfile} = profileSlice.actions;
export default profileSlice.reducer;
