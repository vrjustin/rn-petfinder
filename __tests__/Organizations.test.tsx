import React from 'react';
import {Linking} from 'react-native';
import renderer, {act} from 'react-test-renderer';
import {waitFor, render, fireEvent} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import store from '../stores/store';
import apiService from '../services/apiService';
import * as organizationsReducer from '../reducers/organizationsReducer';
import Organizations from '../Views/Organizations';
import {
  mockOrganizationsResultsResponse,
  mockOrganization,
} from '../__mocks__/mocks';

jest.mock('../services/apiService.ts', () => ({
  getOrganizations: jest.fn(),
}));

jest.mock('react-native/Libraries/Linking/Linking', () => ({
  openURL: jest.fn(),
}));

let renderedOrganizationsTree: any;

describe('Organizations', () => {
  it('renders correctly', async () => {
    (apiService.getOrganizations as jest.Mock).mockResolvedValue(
      mockOrganizationsResultsResponse,
    );
    await act(async () => {
      renderedOrganizationsTree = renderer.create(
        <Provider store={store}>
          <Organizations />
        </Provider>,
      );
    });
    expect(renderedOrganizationsTree.toJSON()).toMatchSnapshot();
  });

  it('handles error when getOrganizations fails', async () => {
    (apiService.getOrganizations as jest.Mock).mockImplementationOnce(() => {
      throw new Error('Mocked Error');
    });
    console.error = jest.fn();

    await act(async () => {
      renderedOrganizationsTree = renderer.create(
        <Provider store={store}>
          <Organizations />
        </Provider>,
      );
    });

    expect(console.error).toHaveBeenCalledWith(
      'Failed to fetch organizations: ',
      expect.any(Error),
    );
  });

  it('handles touching an organization and calling openUrl on item.url', async () => {
    jest
      .spyOn(organizationsReducer, 'selectOrganizations')
      .mockReturnValue([mockOrganization]);

    const {getByTestId} = render(
      <Provider store={store}>
        <Organizations />
      </Provider>,
    );

    const orgButton = getByTestId(`Organizations-Org-${mockOrganization.id}`);
    fireEvent.press(orgButton);

    const expectedURL = mockOrganization.url;
    await Promise.resolve();
    expect(Linking.openURL).toHaveBeenCalledWith(expectedURL);
  });
});
