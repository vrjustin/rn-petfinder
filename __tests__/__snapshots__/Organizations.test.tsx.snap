// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organizations renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 0,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 32,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": undefined,
              "fontSize": 30,
            },
            {
              "marginRight": 8,
            },
            {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      style={
        {
          "marginHorizontal": 16,
        }
      }
    >
      3
       / 
      5
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": undefined,
              "fontSize": 30,
            },
            {
              "marginRight": 8,
            },
            {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "_links": {
            "animals": {
              "href": "href",
            },
            "self": {
              "href": "href",
            },
          },
          "address": {
            "address1": "123",
            "address2": "456",
            "city": "Beverly Hills",
            "country": "USA",
            "postcode": "90210",
            "state": "CA",
          },
          "adoption": {
            "policy": null,
            "url": null,
          },
          "distance": 10,
          "email": "testorg@testemail.com",
          "hours": {
            "friday": null,
            "monday": null,
            "saturday": null,
            "sunday": null,
            "thursday": null,
            "tuesday": null,
            "wednesday": null,
          },
          "id": "1",
          "mission_statement": null,
          "name": "TestOrgName",
          "phone": "555-555-5555",
          "photos": [
            {
              "full": "full",
              "large": "large",
              "medium": "medium",
              "small": "small",
            },
          ],
          "social_media": {
            "facebook": null,
            "instagram": null,
            "pinterest": null,
            "twitter": null,
            "youtube": null,
          },
          "url": "www.url.com",
          "website": "www.website.com",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "gray",
              "borderRadius": 5,
              "flexDirection": "row",
              "marginBottom": 10,
              "marginHorizontal": 16,
              "opacity": 1,
              "padding": 10,
            }
          }
          testID="Organizations-Org-1"
        >
          <View
            style={
              {
                "marginRight": 10,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "white",
                    "fontSize": 20,
                  },
                  {
                    "color": "white",
                    "fontSize": 20,
                  },
                  {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View>
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                TestOrgName
              </Text>
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 14,
                  }
                }
              >
                Beverly Hills, CA
              </Text>
            </View>
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 12,
                  "marginLeft": "auto",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
